stdlib/public/SDK/Foundation/Data.swift

```diff
@@ -1244,35 +2049,71 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
+            if providesConcreteBacking {
+                _representation = _Representation(_DataStorage(immutableReference: reference.copy() as! NSData, offset: 0), count: length)
+            } else {
+                _representation = _Representation(_DataStorage(customReference: reference.copy() as! NSData, offset: 0), count: length)
+            }
         }
         
     }
     
     // slightly faster paths for common sequences
     @inlinable
-    public init<S: Sequence>(_ elements: S) where S.Iterator.Element == UInt8 {
-        let backing = _DataStorage(capacity: Swift.max(elements.underestimatedCount, 1))
-        var (iter, endIndex) = elements._copyContents(initializing: UnsafeMutableBufferPointer(start: backing._bytes?.bindMemory(to: UInt8.self, capacity: backing._capacity), count: backing._capacity))
-        backing._length = endIndex
-        while var element = iter.next() {
-            backing.append(&element, length: 1)
+    @inline(__always)
+    public init<S: Sequence>(_ elements: S) where S.Element == UInt8 {
+        // If the sequence is already contiguous, access the underlying raw memory directly.
+        if let contiguous = elements as? ContiguousBytes {
+            _representation = contiguous.withUnsafeBytes { return _Representation($0) }
+            return
+        }
+
+        // The sequence might still be able to provide direct access to typed memory.
+        // NOTE: It's safe to do this because we're already guarding on S's element as `UInt8`. This would not be safe on arbitrary sequences.
+        let representation = elements.withContiguousStorageIfAvailable {
+            return _Representation(UnsafeRawBufferPointer($0))
+        }
+
+        if let representation = representation {
+            _representation = representation
+        } else {
+            // Dummy assignment so we can capture below.
+            _representation = _Representation(capacity: 0)
+
+            // Copy as much as we can in one shot from the sequence.
+            let underestimatedCount = Swift.max(elements.underestimatedCount, 1)
+            _withStackOrHeapBuffer(underestimatedCount) { (buffer) in
+                // In order to copy from the sequence, we have to bind the buffer to UInt8.
+                // This is safe since we'll copy out of this buffer as raw memory later.
+                let capacity = buffer.pointee.capacity
+                let base = buffer.pointee.memory.bindMemory(to: UInt8.self, capacity: capacity)
+                var (iter, endIndex) = elements._copyContents(initializing: UnsafeMutableBufferPointer(start: base, count: capacity))
+
+                // Copy the contents of buffer...
+                _representation = _Representation(UnsafeRawBufferPointer(start: base, count: endIndex))
+
+                // ... and append the rest byte-wise.
+                while let element = iter.next() {
+                    Swift.withUnsafeBytes(of: element) {
```

Comment:
This looks like a terrible performance pitfall to me... No benchmarks currently cover this, but in my local test, this is 32x slower than hitting the happy `_copyContents` path.

Reply:
Sorry for the delay in responding — just got back from vacation.
Just to clarify, do you mean by this comment that we should slurp all of the elements out of the iterator into an array, then append to `_representation`, or just that hitting this code path is slow? (Because the latter is somewhat expected — we've got a sequence here which is not `ContiguousBytes`, doesn't have an implementation of `withContiguousStorageIfAvailable`, _and_ doesn't implement the `_copyContents` optimization, AKA, the worst-case scenario.)

Reply:
The first. I don't think such slowdown is acceptable. Here's benchmark I got:

````swift
…
  BenchmarkInfo(name: "Data.append.Sequence.ExactCount",
    runFunction: { append($0*100,
      sequence: repeatElement(UInt8(0xA0), count: 809), to: medium) }, tags: d),
  BenchmarkInfo(name: "Data.append.Sequence.UnderestimatedCount",
    runFunction: { append($0*100,
      sequence: repeatElementSeq(809), to: medium) }, tags: d),

…

let repeatElementSeq = { count in
  return sequence(state: count) { (i: inout Int) -> UInt8? in
    defer { i = i &- 1 }; return i > 0 ? UInt8(0xA0) : nil
  }
}

@inline(never)
func append<S: Sequence>(_ N: Int, sequence: S, to data: Data)
  where S.Element == UInt8 {
  for _ in 1...N {
    var copy = data
    copy.append(contentsOf: sequence)
  }
}
````

Reply:
...and just to clarify, we probably don't need to slurp **all of the elements** out of the iterator in one go, just amortize the costs of going over the `UnsafeBufferPointer` by slurping more elements in batches periodically append to the `_representation`. We could try reusing `InlineData`, or test some moderately sized `ContiguousArray`…

Reply:
@palimondo Sure, this is a reasonable change we can and should make. Are you interested in taking this on?

(We should make a similar change in `append<S : Sequence>(_: S)` as well.)

Reply:
Not at the moment. I want to first finish my cleanup pass of the Swift Benchmark Suite, applying the `legacyFactor`.

Also, when I have played with this locally (I was `println` debugging `Data` implementation to see which branch get's exercised), I had trouble getting the build to work properly. Maybe it was just that most of the impl was inlined and the DataBenchmarks were't getting rebuilt? It was a mess.

So I cannot promise I could get back to this sooner than in two weeks...

Reply:
@palimondo No worries, I'll take care of this, possibly in #21754